File: src/algorithms/Minimum_Window_Substring.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {string} s
    # @param {string} t
    # @return {string}
    def minWindow(self, s, t):
        s_rec, t_rec = {}, {}
        st, i, ret = 0, 0, None
        for ch in t:
            if ch not in s_rec:
                s_rec[ch] = 0
            if ch not in t_rec:
                t_rec[ch] = 1
            else:
                t_rec[ch] += 1
        while st < len(s) and i < len(s):
            if s[i] in s_rec:
                s_rec[s[i]] += 1
            tag = True
            for key in t_rec:
                if s_rec[key] < t_rec[key]:
                    tag = False
                    break
            if tag:
                while (s[st] not in t_rec or
                       s_rec[s[st]] > t_rec[s[st]]):
                    if (s[st] in s_rec and
                       s_rec[s[st]] > t_rec[s[st]]):
                        s_rec[s[st]] -= 1
                    st += 1
                if ret:
                    if ret[1] - ret[0] > i + 1 - st:
                        ret = (st, i + 1)
                else:
                    ret = (st, i + 1)
            i += 1
        return s[ret[0]:ret[1]] if ret else ""


if __name__ == '__main__':
    solution = Solution()
    print solution.minWindow('ADOBECODEBANC', 'ABC')

File: src/algorithms/cpp/Minimum_Window_Substring.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    string minWindow(string s, string t) {
        bool vis[256] = {};
        map <char, int> s_rec, t_rec;
        int l_pos = 0, cnt = 0;
        int ret_l = -1, ret_r = -1, minn = SZ(s) + 1;
        for (int i = 0; i < SZ(t); ++ i)
            ++ t_rec[t[i]];
        for (int i = 0; i < SZ(s); ++ i) {
            ++ s_rec[s[i]];
            bool tag1 = false;
            if (s_rec[s[i]] >= t_rec[s[i]]) {
                if (!vis[int(s[i])])
                    vis[int(s[i])] = true, ++ cnt;
                if (cnt >= SZ(t_rec))
                    tag1 = true;
            }
            if (tag1) {
                while (l_pos <= i) {
                    if (s_rec[s[l_pos]] > t_rec[s[l_pos]]) 
                        -- s_rec[s[l_pos]], ++ l_pos;
                    else 
                        break;
                }
                if (i - l_pos + 1 < minn) {
                    ret_l = l_pos, ret_r = i + 1;
                    minn = i - l_pos + 1;
                }
            }
        }
        if (ret_l == -1)    return "";
        else                return string (s.begin() + ret_l, s.begin() + ret_r);
    }
};

int main() {
    Solution solution = Solution();
    cout << solution.minWindow("ADOBECODEBANC", "ABC") << endl;
    cout << solution.minWindow("a", "a") << endl;
    return 0;
}
